### DISCUSSION TITLE
Does spdlog support utf-8 encoding in Linux?

### DISCUSSION THREAD

**Discussion by RonnieHu** [gabime/spdlog#2462] [Q&A] [answered]
I want to write the log file in form of utf-8 encoding.
And I hava tried the "SPDLOG_WCHAR_TO_UTF8_SUPPORT" macro, which returns the error of "#error SPDLOG_WCHAR_TO_UTF8_SUPPORT only supported on windows"
So is there some way to support the utf-8 encoding style in spdlog?
Thanks

**Comment by tt4g** [comment, marked as answer]
Short answer: spdlog does not support `char *` character encoding conversion.

Long answer: In the C/C++ language, there is no character encoding specification for `char *`. The lack of specification means that it is compiler and platform (OS) dependent.
Since there is no language specification, there is no way to know spdlog the character encoding of the data assigned to `char *`.

However, most Linux OS treat `char *` in source code as UTF-8 encoding.
More strictly, the GCC compiler uses the `exec-charset` option to determine the character encoding of `char *` in the source code (https://gcc.gnu.org/onlinedocs/cpp/Invocation.html).
It is safe to assume that the character encoding used by the compiler and platform is UTF-8 from the beginning.

NOTE: actual applications read external files and exchange data via HTTP request/response.
The encoding of these strings must be converted to the character encoding in the application for which the developer is responsible.

**Reply by tt4g** [reply]
See also: https://stackoverflow.com/questions/148403/utf8-to-from-wide-char-conversion-in-stl

**Comment by tikabass** [comment]
As of c++23, there will be native UTF-8 support in c++. Are there any plans to support c++23's char8_t ?

**Reply by tt4g** [reply]
The `char8_t` type is just to make it easy to tell that a UTF-8 string is being stored.  There is no magic that will convert the encoding of the assigned character to UTF-8.
The data size is the same as `char`, so it can be safely cast with `static_cast<char *>(char8_t *)`.

Reference: https://en.cppreference.com/w/cpp/language/types#char8_t
